### ISSUE TITLE
'Relative path' too strict - 'ep' profiles fail when copying nested directories to pool (drive(s) nearly full)

### ISSUE THREAD

**Issue by malventano** [trapexit/mergerfs#757]
Botton line up front:
As the 'ep' profiles work currently (or at least epff/eplfs/eprand), the user would need to manually intervene constantly if some drives with the relative paths were nearly full and the incoming content had nested directories. Switching away from an existing path profile may not be an acceptable workaround for those who were after the 'ep' functionality for their use case (filling drives sequentially, etc).

Proposed solution:
'ep' profiles should behave less strictly when all relative path drive(s) are full, where the code 'walks back' the path and broadens the drive search until a successful path is found (this would maintain the goal of grouping to within drives that had some portion of that relative path tree). This can be done either with new profiles or changing the existing 'ep' profile behaviors with a possible 'strict ep' flag for those users needing a more strict existing path behavior.

Example config:
category.create=epff
ignorepponrename=true
moveonenospc=true
minfreespace=50G
use_ino
cache.files=off
dropcacheonclose=true
allow_other

Example scenario:
Several drives of the pool contain the /foo/ path and of those, some are just shy of the minfreespace setpoint. Now attempt copy in a new set of files/folders with the following tree:
/bla1/
/bla1/test1/ (some files here)
/bla1/test2/ (some files here)
/bla1/test3/ (some files here)

During this copy, /foo/bla1 is created on a drive with the ep, then /foo/bla1/test1/ is created on that same drive, but during the copy, drive 1 reaches its minfreespace setpoint. Since /foo/bla1/test1/ and /foo/bla1/test1/ does not exist on any other drives, we are now at a deadlock where the folder tree can be completely/successfully copied to the pool without manually creating /foo/bla1/test1/ on another drive (/foo/bla1/ is not sufficient). Even after this is done, if /foo/bla1/test2/ fills the next drive to its minfreespace limit, we have to do this whole manual intervention again. Repeat ad nausea. More frustratingly, even if the user manually creates the nested folders on other drives and reattempts the copy, some smaller files contained within subdirs nested further down may cause those dirs to be created on the first drive, but then the rest of the contents can't fit on the second drive as the path doesn't go that deep on that drive.

**Comment by trapexit** [comment]
This would significantly increase the amount of system calls required for such actions and would ultimately undermine the point of keeping data segregated. If you walk back you eventually hit the root. Meaning you'll just clone the full path to /y/a/b/c/d/e/f to /z/a/b/c/d/e/f eventually. And as I ask / point out in the docs... why bother with colocation at all at that point? There are few reasons to colocate data and as far as I know of the use cases there are combinations of policies that work just fine for those people.

**Comment by malventano** [comment]
The increased system calls would only occur for a new nested folder being added and only when the existing path was out of free space. The next file written to that path would go to the next drive (which had its full path created on the prior call). The remainder of the copy would also complete as expected since more of the tree now exists on the next drive.

Alternative pooling options I have tested (StableBit DrivePool) are able to be similarly configured (fill drives sequentially) and can handle this exact scenario without drive full errors, but there appears to be no way to get MergerFS to function similarly.

If there is some existing configuration that will allow the use of epff and copying a set of nested folder content over and have it succeed while crossing over the point where one of the drives filled, I'd be happy to test it, but at present this is an issue that users of the epxx profiles are likely to run into, and would likely just switch to a non-ep profile out of frustration. epff should be able to handle copying a set of nested folders without drive full errors.

**Comment by trapexit** [comment]
>The increased system calls would only occur for a new nested folder being added and only when the existing path was out of free space.

So you're saying it shouldn't walk the tree normally. Just when out minfreespace is hit? When zero branches are eligible recursively scan all branches till a shared relative path is found and clone the path from one of the paths that did exist (cause it could be N)?

>Alternative pooling options I have tested (StableBit DrivePool) are able to be similarly configured (fill drives sequentially) and can handle this exact scenario without drive full errors, but there appears to be no way to get MergerFS to function similarly (while also retaining ep function of MerferFS).

How does "fill drive sequentially" and keeping paths work? Filling a drive sequentially means you don't care where random data ends up you just want to fill one drive at a time for some reason. Path preservation is for those who care where data ends up.

>If system calls is the issue, is it possible to use epff but fallback to ff on drive full errors? That would resolve the problem on the next file write and then epff could continue as normal.

If you ignore path preservation what is the point of having it on at all? You will silently increase the layout entropy over time and any reasons you had to segregate data get complicated as you track down where the split paths are.

 >The reason for colocation was in the OP - reads of a folder of data should come from a minimal number of drives, as to prevent multiple drives from spinning up on a given folder copy).

Why would that minimize drive spin up? The mere act of querying a drive will usually cause a spin up. Even if the files full paths were known and the software wasn't searching the directory mergerfs still has to find the file which will result in some scanning to find it.

**Comment by malventano** [comment]
> So you're saying it shouldn't walk the tree normally. Just when out minfreespace is hit? When zero branches are eligible recursively scan all branches till a shared relative path is found and clone the path from one of the paths that did exist (cause it could be N)?

Exactly.

> If you ignore path preservation what is the point of having it on at all?

The point was not to ignore it, but to workaround the copying of nested dirs issue while retaining ep functionality. Walking back the path on a drive full event would enable the copied tree to remain within the drives containing the destination path (/foo/) and not spill over to the drives not containing that path unless all drives containing /foo/ were full.

Combining the two above, maybe there's merit to setting a depth to which this is allowed to walk back e.g. set to a depth of 1, walking back will stop at /foo/ and then if all drives containing /foo/ are full, then report drive full instead of spilling over onto drives not intended for that data.

> How does "fill drive sequentially" and keeping paths work?

Hopefully with the above, we can make it work :)

> Why would that minimize drive spin up?

When merging over a block cached file system, all other drives stay asleep. I've confirmed this with ZFS volumes. The directory data remains in the ARC/L2ARC. It's way better at this than I was expecting it to be, and even handles repetitive plex media scans of the pool without spinning up any drives.

**Comment by malventano** [comment]
Re: setting a depth as a config option:

This could be the way to enable the workaround we're describing here - if the user sets a config value then drive full events will walk back as far as the user-selected depth (building up relative to root, not back from the current path). if not set, then the default is the prior behavior e.g. no walkback.

**Comment by trapexit** [comment]
>The point was not to ignore it, but to workaround the copying of nested dirs issue while retaining ep functionality. 

Well... that's a redefinition of what path preservation means. It's not preserving paths... it's more like sticky paths. Prioritize what exists. That's fine... though I'm not sure how different it is in practice to using mkdir=rand or mkdir=epall which others use. Any sufficiently complex system will have a number of access patterns. Optimizing for any one of them rarely works out in the long run. Hence why most people give up on trying to colocate.

>When merging over a block cached file system, all other drives stay asleep. I've confirmed this with ZFS volumes. The directory data remains in the ARC/L2ARC. It's way better at this than I was expecting it to be, and even handles repetitive plex media scans of the pool without spinning up any drives.

You're literally the only person I've heard of accomplishing that. Few people however have such a setup. Regardless, I'm not sure I understand why you want colocation / path preservation then. If it caches all the data such that even plex scans aren't triggering a spinup... what does path preservation buy you?

**Comment by trapexit** [comment]
I'm asking all these questions because *many many* people have asked about path preservation over the years and 9 times out of 10 they don't actually need it or think it would give them something it doesn't. I'm just trying to make sure before anyone goes off to write code that the whole of the situation and ask is understood.

**Comment by malventano** [comment]
Totally understand.

Regarding the caching thing: yeah, I was surprised that ZFS ended up so effective and dead easy (compared with the alternative block cache solutions that I know you've dabbled with). I haven't even enabled L2ARC yet and just with ARC it's *really* effective at minimizing spinup (so long as atime=off). The pauses come only when a drive spins up for access to uncached data, but what you want to avoid is multiplying that wait out across multiple drives, which will happen as data within folders gets spread across several drives over time (when using a non-ep profile). Without ep in effect, you'd end up where you copy some folder of media files and run into repeated waits for spinup as each new file was located on some other drive. 'ff' gets you close but then you lose all of the ep advantages (next paragraph) and also in the case where some space was freed on the first few drives, a new single folder of content will fill those gaps and then be spread across those drives, resulting in multiple serial spinups when that same folder is copied elsewhere in the future.

There are many advantages to path preservation with existing paths preset. A user can have access to a common pool containing HDDs, SSDs, different sets of data that are meant to be segregated in various ways on the backend. Sticking with my example, let's assume /foo/ was on one set of disks and was not vital data, but /vital/ was on another set of disks. SnapRAID could be running on the volumes containing /vital/ while not touching the (perhaps older/slower) /foo/ disks. SSDs could be present within the same large pool, all with /ssd/ in their root, and we'd be able to copy nested dirs into /ssd/ and know they will fill the SSDs and spill over onto the next drive of that path as each fills (including large files - see * below - I found a related edge case pertaining to moveonenospc=true combined with ep profiles). Bluntly put - stuff just gets spread all over the place when not using epxx and even ff doesn't win here :).

- A separate thought process from my use case / the example here, but I was going to expand and say that pool-based SSD caching scenarios are also related, but would require a different solution. If a caching SSD is the first drive in a pool and using epff to ensure incoming data goes to that SSD first, the docs mention the user would need to make the relative path on the SSD, but the reality is that the user would need to recreate the entire tree of the pool on that SSD, otherwise subdir operations would end up going to the HDDs that contained those deeper paths, bypassing the cache. Again, a separate issue but there might be some sort of overlap with the increase in flexibility related to walking back paths, etc.

* As I've been experimenting here, I did notice that moveonenospc=true is in a roundabout way duplicating paths onto another drive, but not following epxx when it does so. If the file that takes the drive past minfreespace also overfills the drive, then MergerFS duplicates that full path elsewhere (it currently appears to follow 'mfs' for this), even if that drive has zero dirs present. It did this with epff at play even though there were several other drives with plenty of space and with the majority of that path present. What should be happening is the same 'walk back the path' operation as to find the next drive with most of the path present (that can fit the file). Instead, it moved that file to a completely empty drive, and such an empty drive in a pool would end up accruing the random overfilled file paths where those should be staying on drives that had more of the tree present.

...so given all of this, ideally we'd want to walk back the path (to a user-set threshold) when using epxx profiles, and we'd want to do this both for cases where the full path doesn't exist (to cover the nested folder copy case), as well as for actions taken for moveonenospc=true. I believe such a solution would satisfy most of those '_many many_' people who might have wanted to do it one way over another, so even if they didn't actually need it at least you don't have to convince them to do it another way. It would add some flexibility to MergerFS, increase its strengths (epxx profiles have less edge case full space error scenarios), and even add functionality (setting the depth when 'walking back').

> sticky paths

hmm... 'spff'? :)

**Comment by malventano** [comment]
...to expand on the SSD caching tangent above, and to explain how my situation leads me to copy large chunks of nested folders into an epff array and run into this drive-full case so easily/frequently, here's my current method (simplified for description). Note that the same applies to regular use of an epff pool so long as the user copies nested folders onto it with a drive nearly full:

'ssd' is a raidz2 to offer fast writes / some fail protection for new data
hdd1-3 are data disks (zfs single-drive volumes)
hdd4-5 are parity (via SnapRAID)

- The samba shared array is ssd:hdd1:hdd2:hdd3 using ff (not epff). All 'new' data goes to the ssd path and worst-case spills over onto hdd1, but this shouldn't happen too often. Call this the 'ff pool'.
- A second (unshared) MergerFS pool is hdd1:hdd2:hdd3 epff (without the ssd). This pool is used for the 'check-in' batch operation noted below. Call this the 'epff pool'
- hdd1-3 is protected by parity of hdd4-5
- Incoming data goes to ssd
- Deleted files go to /.recycle/ per samba config (and stay on their drive as ignorepponrename=true, so those files are still available should SnapRAID need them for rebuilding)

The check-in/maintenance process migrates the ssd data onto the hdd array periodically / once ssd fills to a threshold:
- wake all drives
- copy ssd contents to the epff pool (this triggers the drive full issues noted above)
- trigger a SnapRAID sync (updates the parity for the newly added data)
- perform hdd smart logging/maintenance
- if all was successful, delete the duplicate files (present on epff pool) from the ssd, leaving a single copy of all on the ff pool
- (optional) purge /.recycle/ on the epff pool
- re-enable hdd spindown
- various zfs snapshots are taken during this process to protect the dataset against ransomware or bad users on the Windows boxes accessing the share. didn't want to make the above more involved than necessary, but the idea is to 'lock in' the state of the files related to the SnapRAID sync as to prevent modification to lots of files between syncs mangling enough parity as to make recovery less possible.

**Comment by trapexit** [comment]
For me many of your usecases are achievable via other, usually more explicit means, so I wouldn't consider them an advantage necessarily. For instance: users who have different needs for their data often will just create multiple mounts. Functionally there isn't much difference between one mount and two preserved path directories and two mounts to those directories. 

>otherwise subdir operations would end up going to the HDDs that contained those deeper paths, bypassing the cache.

That depends on the mkdir policy. If you use epall for mkdir and epff for create and put the SSD first then it'd work fine. Besides... I don't think anyone using that model is using path preservation. Very few users use path preservation relatively speaking.

> moveonenospc

That was added long ago and in my experience not something you'd want with path preservation and is largely irrelevant when using mfs (the most common policy.) 

>solution would satisfy most of those 'many many' people who might have wanted to do it one way over another

I doubt it will make much difference. They would still need to understand the depth logic. Most literally just "like" the idea of colocating the data without considering all the consequences or due to misunderstanding what it will give them.

>hmm... 'spff'? :)

Not sure how I'd want to implement all this. Creating 'sp*' would result in needing a whole set of new policies. If it can be a optional part of existing policies might be easier.

**Comment by malventano** [comment]
> That depends on the mkdir policy. If you use epall for mkdir and epff for create

Agreed, but it would also wake all drives every time a directory was created as opposed to just waking the next drive that was about to be filled.

I suspect the simplest workaround would be to allow the user to choose the policy used for drive full / moveonenospc events (or simply revert to the non-ep variant in those cases). It wouldn't be as 'cool' as walking back the paths, but it would at least prevent the no space issues and allow the use of epff without the need to spam the full directory tree across all drives.

**Comment by trapexit** [comment]
>Agreed, but it would also wake all drives every time a directory was created as opposed to just waking the next drive that was about to be filled.

That's not guaranteed. And the degree depends on the setup. Again, most people don't have a setup with a large cache in front of drives. 99%+ of users have just a pool of drives. Just scanning the drives to see what paths exist will often spin up drives. For most people they 1) don't even need colocation. They literally have no need. They just like the idea. 2) For those that want it anyway using random mkdir gives them more or less what they think they want. 

You've got a rather unique setup and that needs to be considered.

> (or simply revert to the non-ep variant in those cases)

Ignoring path preservation ever destroys the whole usecase for most users. Years ago it did just that. It was not done on purpose. What happened was that people lost track of what went where and most realized they didn't need it in the first place. It was a feeling. It served no other purpose.

**Comment by malventano** [comment]
This is sounding more like your stance is that 'nobody should ever use epff'. I've pointed out that there are use cases where it's beneficial, and clearly some subset of folks like the ep policies, so shouldn't these policies not fail every time an additional file is copied to a path down the tree that exists on just a single drive? Especially if we have a case where a large file might just go and follow a non-configurable policy that may be the opposite of what the user wants? If anything it's inconsistent now in that with the OP settings, smaller files being added down a particular path will result in drive full errors, but if you just happened to write a large enough file at just the right time, then the path was duplicated on another drive and the error was averted. The ideas I'm proposing here addresses all of that without the need to just give up on ep profile usage.

**Comment by malventano** [comment]
> That's not guaranteed. And the degree depends on the setup.

If you always make the directories on all drives, then every new directory created will wake all drives. Even with caching. To avoid it would require infinite writeback timeout, which is dangerous at best.

Separately, regarding epall for mkdir, is there not lots of overhead being added in the form of the creation of each dir on all drives, as well as looking for files down paths that are empty on most drives? Given none of the other mkdir policies will avoid the full errors here, it just seems like a brute force / messy option as compared to following a slightly different behavior on those rare drive full events.
I get that you want the ep to be 'strict', and it still can be 99.99% of the time, but it's useless if that random 0.01% where you filled a drive results in a deadlock where the user can't add more files to the folder without micromanaging path creation or hitting it with a rather large blunt setting/workaround.

'Existing Path' just doesn't need to be *that* strict. The user should be able to copy in a tree of dirs to a path off of the root that existed across multiple drives, without hitting drive full errors. If you believe that most people don't need ep policies, then since only a minority are using them, wouldn't we want those folks who are using them (regardless of reason) to do so without tripping these error conditions?

**Comment by trapexit** [comment]
I've not said that and I'm not saying that. I'm saying that most people who claim to want the feature in fact have no use for it. As I pointed out in this thread and in the docs.

>so shouldn't these policies not fail every time an additional file is copied to a path down the tree that exists on just a single drive?

It depends on your expectations / needs. The intent of path preservation was to do just that. If you are looking just kinda preserve paths to get some data locality then in *most* cases the person in fact has no need for the features. Most people think it will keep drives from spinning up. When I said you're literally the first person to claim to "solve" the drive spinup problem I meant it. So for those hundreds of people who have pinged me about it... they were incorrect because their setup would never allow for that without massive caching on mergerfs' side which I'm not interested in implementing given the niche usecase. The only people who have had legit needs for path preservation where people without backup of any kind and wanted to lose data in logical chunks (TV seasons, music albums, etc.) or who liked to take drives out of their system regularly for misc reasons and wanted whole tv shows to be available on one drive. For those use cases mkdir=rand works perfectly fine. Those with zfs + mergerfs have never had it setup in a way where any of this would matter. You are the one and only person I've ever seen with that setup and requested behavior. Other's have gotten what they wanted by other means.

>Especially if we have a case where a large file might just go and follow a non-configurable policy that may be the opposite of what the user wants?

moveonenospc was never intended to be used with path preservation. It existed before policies. It is mostly intended as an "oh shit" catch that happens once in a blue moon.

>The ideas I'm proposing here addresses all of that without the need to just give up on ep profile usage.

These ideas aren't new. Every policy under the sun has been proposed or debated but they've just never been necessary. For most people if you're willing to allow a directory of files split across multiple drives it's functionally the same as no colocation at all. Either spinning up of drives happens anyway or you can just put filters in your snapraid conf or whatnot.

**Comment by trapexit** [comment]
>If you always make the directories on all drives, then every new directory created will wake all drives. Even with caching. To avoid it would require infinite writeback timeout, which is dangerous at best. 

As I said it depends on the setup. You're talking about *your* setup. Not most people's. Not mine. For most people there is no caching. The drive will be spinning anyway. So the policy doesn't matter. It won't help keep drives from spinning.

>Separately, regarding epall for mkdir, is there not lots of overhead being added in the form of the creation of each dir on all drives, as well as looking for files down paths that are empty on most drives?

No. It's not all drives. It's all drives with that base path. Why would that be a lot of overhead? You make directories pretty rarely in relation to all other actions. And two mkdirs after some stats is nothing in comparison to most other actions.

 > Given none of the other mkdir policies will avoid the full errors here, it just seems like a brute force / messy option as compared to following a slightly different behavior on those rare drive full events.

I don't follow. Drive full as in minfreespace or getting an actual out of space error? minfreespace can be hit all the time on people using ff or lfs. And almost never when using mfs or lus. Few people use ff on a normal pull of drives. It degrades performance and increases risk of total data loss. Regardless, you have to scan all branches for lots of reasons all the time.

As for policy reported space errors... how do they not avoid it? The only difference between the "walk the tree to some max depth" and "epall" is that "epall" results in some more directories and file spread. The number of files on the drives are the same. And with "rand" if you have any amount of consistency in the amount of data per directory it works perfectly fine. For most hierarchical data sets (TV, Movies, Music) it works fine.

> 'Existing Path' just doesn't need to be that strict. The user should be able to copy in a tree of dirs to a path off of the root that existed across multiple drives, without hitting drive full errors. If you believe that most people don't need ep policies, then since only a minority are using them, wouldn't we want those folks who are using them (regardless of reason) to do so without tripping these error conditions?

Those people aren't running into those problems. If it was a wide scale problem it would have been addressed already. EP policies have been around for years. It's the default behavior. It's widely discussed in the docs. It's hardly a untouched topic.

I'm not opposed to adding features. I'm just saying you're asking for something that up to this point in time has not been needed. Mostly by people who incorrectly thought they did and/or thought it worked a way it did not. Hence being thorough in getting all your details to ensure you are not one of them.

**Comment by malventano** [comment]
> For those use cases mkdir=rand works perfectly fine. 

Just to be sure I understand this correctly, would that not create the directory just once, on a random drive? What happens if that random drive was nearly full and the directory contents take it past the threshold. Do we not get the errors in that case?

> moveonenospc was never intended to be used with path preservation.

This feature should properly function across all profiles, especially those employing any form of 'ff'. I'd even argue it should be a default. Why on earth would a user ever *not* want this to happen? Else you end up with oddball issues depending not only on if a drive became full, but how it did so (growing an existing file, etc).

> Either spinning up of drives happens anyway or you can just put filters in your snapraid conf or whatnot.

Right, but I've demonstrated that the spinup issue can be solved, and snapraid filters don't help in the context of keeping contents on particular drives. That solution also breaks where filters were at play but the unfiltered content, if not following 'ep', end up on drives that might be larger than the snapraid parity drives.

> The intent of path preservation was to do just that.

Ok, let's try and understand the need for path preservation. Put all of the spindown stuff aside. Let's imagine a user wants a few specific drives to contain torrents. They create /torrent/ on a drive and start downloading some torrents. Then they realize they want to add a second drive for that purpose, so they create /torrent/ on an additional drive. Now their bittorrent client, adding more data from a pre-existing torent that was on the first drive fills the drive while writing to /torrent/best.music.ever.collection/album1/ . Now we're stuck, and every additional file the torent client attempts to add to that path fails, while the other drive with /torrent/ has plenty of space. If the torrent client was growing an existing file and that maxed out the drive it was on, of course we'd want to move it to another drive (via moveonenospc). Which drive would we want to move it to? ... the other one with /torrent/ present (currently it would go to mfs which might not be that other /torrent/ drive, defeating the purpose).

Given the above example, I'm having a hard time imaging a case where anyone who did want ep would not want the drive full behaviors to at least to try to follow some form of ep when a drive is full.

**Comment by Wintersdark** [comment]
Why wouldn't that user create an independent MergerFS pool just containing the /drive/torrent folders?  They could add additional /drive/torrent folders any time they wanted.  

I think a lot of people miss out on the potential of MergerFS by trying to cram everything into a single mount, not realizing you can have multiple mounts referencing the same disks and even individual folders within those disks.

**Comment by trapexit** [comment]
>Just to be sure I understand this correctly, would that not create the directory just once, on a random drive? What happens if that random drive was nearly full and the directory contents take it past the threshold. Do we not get the errors in that case?

Yes but you have to consider the start state, other policies, etc. If you have 2 drives. Many people start with fresh drives or will balance between drives before adding a new one. If you start at equal footing then a random mkdir over two drives with a epmfs file create policy works perfectly fine 99% of the time. When drive0 hits it's minfreespace limit the statistical likelihood that drive1 is close to that value is very high. If you're starting with a 95% filled drive and 0% filled drive obviously random wouldn't work for unshared basepaths. But again... not a problem for most people. They aren't doing that. If I have a music directory with a 1000 and 3 drives with random mkdir policy... I'm going to have 333 directories per drive. Again, if you have general uniformity in your directory placement which on the whole is true for the average media setup... it's not really different from what you suggest.

>epff would have resulted in only spinning up one (or two if the first overfilled while writing, spanning the path onto a second drive).

How do you know what drive you're copying from? I don't know about you but my access patterns are all over the place and it would never be possible to predict what would be an ideal layout. And again... spinning up drives is the norm. There is no way the average system would spin down drives if you're doing backups and torrents and streaming and what have you. Only if you carefully segregate data locations can you get that. Few people do it. Those who do go to great effort. And when you do you naturally lose the ability to blanket pool data.

>Right, but I've demonstrated that the spinup issue can be solved

FOR YOU. I'm not (and I doubt many others) putting enough SSDs in my 100+TB setup in the hopes that enough data will be caches so as to not spin up the drives. It wouldn't happen. I've too much data being read and written and processed and streamed and backed up and bitrot detected and and and. It would be extra setup with next to zero payoff.

>snapraid filters don't help in the context of keeping contents on particular drives

If you're talking about spinning up of drives that's a separate concern. The point is that you can have specific drives hold specific data and use filters to choose what data you want to include in snapraid.

>end up on drives that might be larger than the snapraid parity drives.

Most people use the largest drive they have for parities or spread across multiple drives. There are ways to manage that.

>Now we're stuck

You're describing a situation that of course would happen but my point is it is not one that often does. I'm talking about what's normal usage. Not edge cases. Most people don't use or need path preservation. You're already discussing step 3 when we haven't understood 0 - 2. Why do they want just torrents on two drives? If they do have such a need why wouldn't they just create a separate pool with no path preservation? What functional benefits would it bring to create directories more judiciously across those two drives? For most people. None. If you focus data on one drive you'll get worse speeds. Eventually entropy increases as torrents come and go and you'll have some even spread. Speeds will be about as good as you can get and data will be spread across both drives. At best you have some data locality. All the files from one torrent on one drive except those times when drive0 hit minfreespace. But what good is that? If you backup that data it's meaningless where it is. If you don't... it's not like you get to pick which files or which drive dies. 

Now expand that to one's whole pool. Entropy increases if you allow any automatic "failover" onto other drives. Whether that's moveonenospc or ignorpponrename or even walking the directory to a lower point in the tree and creating a new fork. If you limit where that fork happens you're in effect creating a separate non-path preserved pool within the larger one. Over time files come and go and you could very well end up with files all over the place. And when you run out of space ... you'll get an out of space error and yet you have xTBs on your other 2 drives. So then maybe you lower the depth one lower till it happens again. Or maybe you buy another drive. Either way entropy has increased. 

>Given the above example, I'm having a hard time imaging a case where anyone who did want ep would not want the drive full behaviors to at least to try to follow some form of ep when a drive is full.

Your example is *not* the norm. It just isn't. The majority of people couldn't care less where data goes. They use mfs for creates and are done with it. They don't need moveonenospc cause it basically never happens. Then probably 90%+ of people who tell me they need to colocate data in fact don't. Similar to how most people who ask about caching don't in fact need it (cause they don't understand it usually.) For those who did need path preservation they simply didn't run into any significant issues with the mkdir policy change.

No one is arguing that such a "walk back to a lower shared directory" isn't useful in some cases. I'm arguing it isn't useful in most existing cases. And as such... I, nor anyone else, has implemented it.

**Comment by trapexit** [comment]
> Why wouldn't that user create an independent MergerFS pool just containing the /drive/torrent folders? They could add additional /drive/torrent folders any time they wanted.

Perhaps on a RAM strapped system it would make more sense to have a single mount though the general RAM usage of mergerfs is pretty low generally. I think the main reason for memory bloat over time is fragmentation though I've not confirmed that.

**Comment by malventano** [comment]
> Your example is not the norm. It just isn't.

Understood, however it is a thing that will happen to virtually anyone using epff/eplfs plus subdirectories. If your argument is that those profiles are not the norm, that's fine, but just because they are rarely used doesn't mean that they should not be able to be used without running into this issue.

> Now expand that to one's whole pool. Entropy increases if you allow any automatic "failover" onto other drives. Whether that's moveonenospc or ignorpponrename or even walking the directory to a lower point in the tree and creating a new fork.

Yes, any type of failover increases entropy, but what's the use if we hit conditions where we become stuck? We can make it better than it is. moveonenospc currently creates the path in what would likely not be even close to the right drive for those using epxx, which is (needlessly) increasing entropy beyond what should have happened given the selected profile. ignorpponrename *reduces* entropy and increases efficiency as files are kept on their respective drives. Walking back the path in the face of those drive-full cases would allow the use of epff/eplfs without running into these other issues (and would keep entropy under control as well).

> Perhaps on a RAM strapped system

Yeah, I have no argument down that path. MergerFS is extremely space-efficient in my experience (so long as you don't go crazy with caching options), and you'd have to be pretty strapped for RAM for it to be a reason for single-mount...

> Why wouldn't that user create an independent MergerFS pool just containing the /drive/torrent folders?

This is a valid alternative, but it reduces a lot of the flexibility that having a larger pool of drives affords, and adds complication when using samba-share-based /.recycle/ or in the case of zfs, accessing the merged /.zfs/snapshot/ dir via shadow_copy2 (MergerFS is freaking magical when you consider how seamlessly this works). It also requires breaking down / reconfiguring the MergerFS share configs just to add another drive for a specific usage when the user could have just added that base folder to another drive as necessary (which can very easily be done via a 'maintenance' samba share one level higher than the mounted drives), meaning it's so flexible that there's no need to even ssh into the server, edit fstab, change samba configs, etc for such operations. Something that can be resolved by simply adding a folder to a drive should be favorable instead of logging into the server, editing fstab, updating shares, etc. It also doesn't address the drive full case I'm bringing up here, since even if a user was to have overlapping mounts, etc - if any of those mounts use any form of epxx, they will run into these issues when any drives become full down a particular path. The only time this wouldn't be the case is if their ep's were just flat subdirs that were only ever filled with files at that root e.g. with no subdirs, and those dirs were created across multiple drives.

Let's back up the 'ep' concept/reasoning a bit. It currently accomplishes two main things:
- It gives the user some sort of control over what goes on which drive.
- It grows folder trees down a minimal number of drives, improving entropy.

Both of these are excellent features and strengths of ep profiles (for those who desire it), but I believe the disconnect comes as a side effect to #2 not being aware of the intent of #1. The user likely only wanted control at or below some specific path length that they created on the desired drives (e.g. just the first level off of root), but #2 is currently unaware of that and is always assuming that the path, regardless of how deep or how it was created, was the *only* desired path, and if it happens to be on only one drive (desired trait per #2), then we can't continue to add any data into any folders branching off of that unique (single-drive) tree once the drive is full.

Multiple birds can be killed with a single stone with just a bit more control over what happens on those drive-full events. The moveonenospc could better follow the intentions of the selected profile (and could even become the default if it worked well enough, no longer needing to be treated as a last ditch effort). ep profiles could work better / be more usable / at least have some means of getting beyond this stuck scenario while keeping entropy low and keeping the files within the set of drives prescribed by the user, with the potential for automatic failover to additional drives should the user set the max walkback at root/0. That way, even as files overfill drives, their folders would be duplicated to a drive with most of that path present until there was no choice but to back all the way to root and branch to a new drive. Users wanting control at root+1 or some other level 'no matter what' could just set that level and once all drives with that base path become full, then you start getting the no space errors. It's far easier to just go in and make a folder than to start reconfiguring the server once some drives become full, and I'd think it would be nicer if MergerFS just handled that case reasonably instead of backing itself into a corner, rendering the ep profiles somewhat nerfed in usability.

> No one is arguing that such a "walk back to a lower shared directory" isn't useful in some cases. I'm arguing it isn't useful in most existing cases. And as such... I, nor anyone else, has implemented it.

I'm with you. I'm just saying that epff is *really* difficult to use in practice without running into the issue I'd hope we can address here. My guess is that this very issue scares folks away from epff. The first time they run into the free space issue they'd just change profiles, but in my case I'm trying to keep that entropy low, so there's extra desire to keep some form of ep at play.

Either way, I'm just trying to help. MergerFS was awesome enough for me to migrate my entire dataset (456TB RAW) over to Linux, and I'm extremely impressed with how mature it is. If we can't come to something that you believe would be workable, I'd probably dig into the source and make the necessary tweaks, but my bigger hope would be to make epff more usable to those who preferred it (even if they didn't truly need it). That and I'm sure you can tweak your own codebase *way* more efficiently than I could fumble around with on a fork :).

**Comment by trapexit** [comment]
>My guess is that this very issue scares folks away from epff. 

But what use would someone have for epff? Talking about people using a feature is useless without understanding the purpose. Currently I know of cache drive setups and remote mounts. Both which are special situations where they include additional settings which impact behavior. General usage of ff / epff from the start would lead to inefficient usage of drive bandwidth and increases risk of data loss by skewing data layout. Maybe I can create a list of implications of different policy usage but these are things people generally don't think of.

>The first time they run into the free space issue they'd just change profiles, 

And/or they ping me and ask about the situation and 99/100 times it turns out they've done something they didn't understand or didn't need.

>but in my case I'm trying to keep that entropy low, so there's extra desire to keep some form of ep at play.

The "keeping entropy low" thing is what I'm not understanding. What does that accomplish? 

>If we can't come to something that you believe would be workable

Again, I'm not saying I wouldn't work on it I'm merely trying to understand how it would be used and why. I'm not kidding or exaggerating when I say that 99% of people who say they want certain policy features have absolutely no need for it. They don't understand it. They will possibly be negatively impacted by it. And I don't want people to go down bad paths where they spend a bunch of time and/or money and get burned in the end. I've gone so far to prove to people they didn't need something by implementing the whole feature and letting them use it to show it didn't do what they expected. My point being that it's not about the work. It's about the function and how valid it is.

**Comment by malventano** [comment]
> The "keeping entropy low" thing is what I'm not understanding. What does that accomplish?

The intent is to keep as close to the 'epff' mentality as possible, as to keep forks off of the tree limited to the fewest number of drives possible, which increases the likelihood that copying a group of files or a 'branch off of the tree' is likely to access a minimum number of drives (in my case for spinup, but there may be other reasons aside from preventing excessive directory creation / overhead caused by the alternative workarounds). With respect to how drives fill, this maximizes the number of drives that remain empty at the tail end of the pool, which increases flexibility should some drives need to be repurposed to another system, or filled by the SnapRAID recovery (which would ideally need an empty drive of the same capacity available to rebuild to, etc). If I copy in the tree per the OP example, I want that tree to go to the drive where that base path existed, but it should be able to gracefully handle the drive becoming full during that process without the need to abandon epff.

> General usage of ff / epff from the start would lead to inefficient usage of drive bandwidth and increases risk of data loss by skewing data layout.

Yes, it would be bandwidth inefficient as you are filling a drive to the end before starting on another for either of the 'ff' profiles, but they are selectable options that the user may desire (whatever the reason), and while alternatives might not have epff, they do have ff. MergerFS is superior here in that it has epff, which gives all of the benefits that the ep profiles provide, but we just need it to be usable without breaking when a drive fills is all.

I'm not seeing how data layout is skewed or data is lost here. If anything the proposed fixes would be keeping the layout *less* skewed (no files going to or paths being created on the 'wrong' drives in the case of moveonenospc). Keeping closer to the drives containing at least a portion of the path on other drive full events would also make the data less skewed, not more. Regardless, it doesn't matter how skewed the data may be when you can't add any more data to a particular path when you have several other drives with plenty of space and most of the path present.

> I've gone so far to prove to people they didn't need something by implementing the whole feature and letting them use it to show it didn't do what they expected.

I can promise you that you wouldn't need to go down this road here as these are actual problems that are easily encountered with the use of epff. While ff would work fine for caching scenarios (heck, I even use it for that as part of my setup, as discussed a few posts back), epff will almost always force the user into the drive full scenario. Just take a set of drives, set epff, and copy more than one drive worth of data (with subdirs) over to it and you will hit this issue. Why even have the option exist if it's so likely to run into an error condition that it becomes effectively unusable? Are the options not there in order to enable the user to configure for their use case and/or preference? Wouldn't it be more beneficial to have the selectable option function better/without erroring in normal usage instead of trying to tell folks 'they are holding(using) it wrong'?

Also, I've seen the moveonenospc=true workaround used by many folks posting their configs in various threads (that's how I learned that fuse.mergerfs is needed for the fstab entry to work, but that's a separate thread). Point being that the things you believe are rarely used 'last ditch efforts' may actually be things that folks just look at and go 'yeah, I prefer that functionality'.

Heck, even after reading through the docs several times, I thought the 'relative path' meant some portion of the path and not that the full path needed to exist on another drive else you hit a full space brick wall. I'd imagine others might come away with the same understanding and then be confused when it acts the way it does currently.

**Comment by trapexit** [comment]
>which increases the likelihood that copying a group of files or a 'branch off of the tree' is likely to access a minimum number of drives

I don't see how this is long term sustainable. If you remove data on drive0 after being filled then it will because the target for new directories in that subtree. Once it fills it will create whatever directories on the drive1. Over time you will end up with subtrees spread across all drives down to the depth limit. Entropy still increases over time. It would be especially true of all the other ep policies. epmfs and eplus would end up looking like mkdir=epall.

>which increases flexibility should some drives need to be repurposed to another system

It just keeps it possibly available vs requires an rsync. It also means on failure you'll have to restore more data or no data. It means you have less throughput because data is more condensed. It means you're running drives without any benefit. Putting wear on them for the off chance you need storage? There are arguable better ways to manage that. I've never heard anyone argue they want to keep drives active and unused.

>I'm not seeing how data layout is skewed or data is lost here.

The data is skewed as in front loaded. As in unevenly spread out. If you have 5 drives and you only fill them one at a time and haven't filled the first drive and you lose the first drive you lose 100% of your data. If you have another policy that spreads data over 5 drives you lose 1/5th your data. You also get better performance on average because the likelihood of one drive being accessed via multiple files is reduced. 5 files on 5 drives at 100MB/s each vs 5 files on 1 drive at 20mb/s each. 

> that are easily encountered with the use of epff.

And *noone* arguing that they don't in those situations. The argument has been that basically no one has that situation.

>Why even have the option exist if it's so likely to run into an error condition that it becomes effectively unusable? 

Because it's not unusable to the people who have or do use it. I've provide those examples.

>erroring in normal usage 

Your definition of "normal usage" is wrong. You are looking at things in one way. That's not reality. You are literally the first person ever to complain about it after tens of thousands of users. Those policies have been niche and as such those users use them in different ways than you are proposing.

> instead of trying to tell folks 'they are holding(using) it wrong'?

Because they literally are. You seem to ignore all my examples. This is getting tiring.

>moveonenospc

Yes... lots of people have that enabled and you know what's also true? Lots of people telling me they think it has to do with things it has nothing to do with. Asking why when all drives fill up due to policy mergerfs doesn't move the file somewhere with space (ignoring the policy).  They ask why mergerfs returns enospc when there is space available even though they set a minfreespace. Thinking those are somehow related. They set moveonenospc when they use mfs when statistically it's almost impossible to hit and if it were hit the likelihood it'd succeed is almost zero because by the point you fill up any one drive it means the others are also filled and unless some data happened to be deleted *right* before it'll copy data around for no reason or fail outright when size checks fail.

You fail to grasp how many times I've had these discussions. How many times I get emails, Twitter DMs, etc. on any and all configurations and misunderstanding. Why the docs are so thorough. Many people don't understand how filesystems work. They don't understand the complexity of things they want to accomplish. People who speak in terms of "file copies" who don't understand there isn't any such thing at the filesystem level. Who argue that it's easy to know when there is a filesystem issue and that mergerfs should act appropriately yet have no understanding of the huge number of errors possible and inconsistencies in underlying filesystems. People who ask why posix permissions fail on non-posix filesystems like ntfs and fat. Yes.... there are plenty of people who "do it wrong." 

> relative path

Relative path means the path relative to the base of the branch as configured. Same as the usage of the term anywhere I know it to be used. Vs an absolute path which would be the fully qualified path. mergerfs is a union of paths. The "relative path" would naturally be relative to those provided. What other language is there to use? If "relative" meant any sub path it would have no functional meaning. Because relative '/' exists on every branch. So if it just backtracked until it found a shared point it would include any drive and negate any premise of path preservation. The only policy you could argue that'd not be true is ff. All others would lay out files differently but wouldn't come close to any concept of preserving paths. epmfs where it walked back to relative root would make no sense.

**Comment by malventano** [comment]
> I don't see how this is long term sustainable. If you remove data on drive0 after being filled then it will because the target for new directories in that subtree. Once it fills it will create whatever directories on the drive1. Over time you will end up with subtrees spread across all drives down to the depth limit. Entropy still increases over time. It would be especially true of all the other ep policies. epmfs and eplus would end up looking like mkdir=epall.

I disagree, as 99.99% of the time the data would just be following the normal epff policy without issue, and the drive full condition would only occur once per each drive becoming full, and maybe once again after the drive might have been emptied a bit and then refilled, but we're still talking about just doing something differently 0.01% or less of the time that would enable the user to actually use epff without it hitting a brick wall on a mostly empty array. Such a low percentage of events, combined with walking back the path, would not be likely to just mash everything together into an epall-looking thing even after extended usage.

> It means you have less throughput because data is more condensed.

The datarate for SnapRAID is generally aggregate throughput limited, meaning it's often better to have fewer full drives if at all possible. At worst, it's a wash as far as recovery / sync times go.

> Asking why when all drives fill up due to policy mergerfs doesn't move the file somewhere with space (ignoring the policy).

If this question keeps being asked, perhaps the issue is with the strictness of the policy and not with the user's choice to use it.

> You fail to grasp how many times I've had these discussions.

I get it. I used to get the same sorts of questions about all sorts of storage related issues (I reviewed SSDs/HDDs for over a decade and now test storage for Intel).

How about this. What is your vision of a valid use case for epff as you envisioned it? What prevents that use case from hitting the disk full issue I've noted? I can't think of a single use case for epff that doesn't potentially hit this issue during normal usage, but I'm happy to be proven wrong.

**Comment by trapexit** [comment]
>I disagree, as 99.99% of the time the data would just be following the normal epff policy without issue,

If you hit the minfreespace limit and remove 50MB then it takes only 50MB to fill it again and force things to the second drive. A normal "download X and move or delete it" would constantly lead to that limit being hit. Over time entropy increases. There is nothing to debate. It would happen. It's the number one reason people give for stopping the use of path preservation. They notice their entropy increasing or needing to increase. It's why I stopped using it shortly after I wrote the software. I wrote path preservation FOR ME. 

>The datarate for SnapRAID is generally aggregate throughput limited,

I'm not talking about SnapRAID. I'm talking about mergerfs. Reading more data from few drives is slower.

>If this question keeps being asked, perhaps the issue is with the strictness of the policy and not with the user's choice to use it.

If you set a minfreespace limit and then expect it to be ignored when you hit that limit... what the hell is the point of the limit? If you have epmfs and a clean slate set of drives and the description clearly indicates that the policy requires seeding of paths and you fill one drive... you never understood or read the policy description. Most of those people DO NOT WANT path preservation. They don't understand what it means. They just want their filesystems merged. That's it. Some do say they want colocated data but have no reasoning for it. Saying something like "it just felt right." And when asked what purpose it'd have they literally have nothing besides "feeling."

>What is your vision of a valid use case for epff as you envisioned it?

I didn't "envision" anything. I made ff because that's what mhddfs did that. That was it. epff was just symmetry. People started using ff/epff for the reasons I already gave above and in the docs. 

>I'm happy to be proven wrong.

Before tagged branches, read only ignoring, etc.

You have local drive for downloads of data and rclone for all data reads. You place data into local, having used epff and seeding the basic paths data would go to. You have a very explicit and specific workflow for downloading data and moving data to a remote store. You have random churn on the local download spot. You have mkdir=rand if you have more than one drive. You rclone some "read" path to cloud so you've duplicated all the local paths. Some people didn't want cloud mount to be written to but some did. How they cloned the paths and such matched that desire. Regardless... there was a usecase. Some of it became irrelevant with new features. People today still sometimes have a similar setup even though there are other ways to do it where order is no longer relevant because you can tag the remote branches as not for creates or whatnot.

**Comment by trapexit** [comment]
Instead of arguing with me about what other people do it would have been a much better use of everyone's time to simply articulate in detail the use case as I requested. Proposing solutions is not a good strategy. I need a user story. A description of the general need. I would have much rather discussed the merits of the user story than have to explain why other people don't seem to need what you're proposing. Could have got some coding in tonight rather than writing a novel on github and getting frustrated. This is my hobby. Not my job. I don't enjoy this back and forth when there are things I have to work on.

I don't think it's unreasonable to want to have concrete reasoning behind a request. Especially given the uniqueness of said request and the decent amount of time it will take to build and test such a feature.

**Comment by malventano** [comment]
> Reading more data from few drives is slower.

That's based on an assumption of how files are being accessed. For a media server where the majority of reads are streaming and not bottlenecked by storage, the case is very different.

> You have local drive for downloads of data and rclone for all data reads. You place data into local, having used epff and seeding the basic paths data would go to. You have a very explicit and specific workflow for downloading data and moving data to a remote store. You have random churn on the local download spot. You have mkdir=rand if you have more than one drive. You rclone some "read" path to cloud so you've duplicated all the local paths. Some people didn't want cloud mount to be written to but some did. How they cloned the paths and such matched that desire. Regardless... there was a usecase. Some of it became irrelevant with new features. People today still sometimes have a similar setup even though there are other ways to do it where order is no longer relevant because you can tag the remote branches as not for creates or whatnot.

Ok, so what you are saying is that the use of epff required a bunch of workaround in order to make it work. What I've proposed is an alternative that will negate the need for so many other workarounds (for epff-style use cases that would potentially fill a drive).

> Instead of arguing with me about what other people do it would have been a much better use of everyone's time to simply articulate in detail the use case as I requested.

Apologies. I'm used to it being the other way around, where unless some solution is proposed, the request is generally ignored. That said, I did articulate the use case pretty early on, but we apparently kept talking past each other about the relevance of the use case and the disadvantages of the possible alternatives. I was simply trying to propose what I imagined was a solution to a problem that was likely causing you heartache in the form of multiple folks running into the drive-full issue (and that appears to have been somewhat confirmed).

Regardless, if you decide to do any tweaking, I'm happy to test/debug as needed.

> This is my hobby. Not my job.

Which are your preferred tip jars out of the many you've listed?

**Comment by trapexit** [comment]
>That's based on an assumption of how files are being accessed.

For *most* users they don't control access patterns (nor could they unless they very very carefully manage what apps hit what drives and through mergerfs that can only be done a few ways). As such, on average, putting more files on fewer drives equates to worse performance. Two streams coming from 1 drive can more than half the performance due to seek times. If you have any slow drives it can especially kill performance. Assuming random load... moar drives moar bandwidth (till hitting bus limits).

>Ok, so what you are saying is that the use of epff required a bunch of workaround in order to make it work. 

No. I'm saying that in a niche usecase it works fine. I said that *other, new features* made epff less relevant for that usecase. I said that explicitly. Before those features existed order and directory existence was the only way to differentiate in a way that worked for the setup. epff was *not* a policy they cared about. They cared about differentiating between local and remote mounts. epff was a means to an end. Not the end.

>That said, I did articulate the use case pretty early on

To me it read as a solution. I hate to use Agile terminology but a user story is written plainly. The pain point or motivation or workflow is described. Ideally without any technical proper nouns or implication to what is to be done. Otherwise there will inevitably be anchoring of he conversation and that's not useful. When a user comes and says "I want to limit spinups, please add caching." or "I want caching so I can saturate my 10Gbps NIC." they have presupposed a lot. 9 out of 10 times as I've discussed they don't realize that their ask isn't practical or isn't going to offer them the solution they think it will. It doesn't tell me what they are actually looking for. A better way to articulate it would be: Electricity is expensive in my area / I have drives that are loud and I'm looking for ways to minimize runtime cost (in $$ and dB). There are lots of solutions to any one problem and sometimes the problem is imaginary. Without understanding the motivations and instead moving straight to how it should be solved just makes everything more complicated.

So when I ask why you want colocation of data I'm looking for something like those in the past: "My family likes to bring TV shows with us on vacation. We don't care what we bring because we like forcing ourselves to rewatch old stuff or picking up a new series. The only thing we care about is having full seasons." From there we can consider the tradeoffs of different solutions that may or may not exist in or around the product.

>Which are your preferred tip jars out of the many you've listed?

The support is appreciated.

>if you decide to do any tweaking

I've a few things on my plate at the moment so I'm not sure when I'll get to this.

**Comment by malventano** [comment]
After some experimentation, the following appears to be a partial workaround. It's not ideal, but it does seem to work well enough to prevent running into free space issues from growing files or attempted copying into branches 'far down the tree' (only on one drive) with an ep profile in use:
```
moveonenospc=true
minfreespace=1k
```
Setting minfreespace so low mitigates the errors which prevent initiating the copy (less of a chance that a drive will have <1KB free at the end of the previous copy), and once the copy begins and some file overfills the one drive that had the path (ep profile), moveonenospc then replicates that branch of the tree on the mfs drive and the copy is allowed to continue as expected.

I do still wish there was a way to keep that moveonenospc redirection targeted at some middle ground that tries to stick to a portion of the existing path drives (ideally walking back the path and selecting based on the current profile).

I also wish there was a way to apply that same scheme to minfreespace as to mitigate the edge case where files can no longer be copied into a particular directory, meaning that once the drive with the ep crosses minfreespace/4GB, the next file copy attempt would be allowed to continue but would be redirected to another drive with most of the path present.

Either way, the above config does prevent the forced user intervention for the cases I've tested, and that was the most frustrating part of the issue.

**Comment by trapexit** [comment]
I changed moveonenospc to take a policy. I still think that will simply destroy the whole premise of path preservation but it's there.

> I also wish there was a way to apply that same scheme to minfreespace as to mitigate the edge case where files can no longer be copied into a particular directory, meaning that once the drive with the ep crosses minfreespace/4GB, the next file copy attempt would be allowed to continue but would be redirected to another drive with most of the path present.

What do you mean? minfreespace is simply a filtering value. Has nothing to do with directories present.

It sounds more like you mean something like "pick the branch with the most shared path." That's not the same as walking back or "fallback" to non-path preservation.

**Comment by trapexit** [comment]
https://github.com/trapexit/mergerfs/tree/policy-mspmfs

How about this? See the new policies.

**Comment by trapexit** [comment]
#784
